vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -191,47 +188,62 @@
      * <p>
      * Examples:
      * <pre>
      * <code>
-     * List.empty().average()              // = None
-     * List.of(1, 2, 3).average()          // = Some(2.0)
-     * List.of(0.1, 0.2, 0.3).average()    // = Some(0.2)
-     * List.of("apple", "pear").average()  // throws
+     * List.empty().average()                       // = None
+     * List.of(1, 2, 3).average()                   // = Some(2.0)
+     * List.of(1.0, 10e100, 2.0, -10e100).average() // = Some(0.75)
+     * List.of("apple", "pear").average()           // throws
      * </code>
      * </pre>
      *
+     * Please note that Java's {@link DoubleStream#average()} uses the
+     * <a href="https://en.wikipedia.org/wiki/Kahan_summation_algorithm">Kahan summation algorithm</a>
+     * (also known as compensated summation), which has known limitations.
+     * <p>
+     * Vavr uses Neumaier's modification of the Kahan algorithm, which yields better results.
+     *
+     * <pre>{@code
+     * // = Optional(0.0) (wrong)
+     * j.u.s.DoubleStream.of(1.0, 10e100, 2.0, -10e100).average()
+     *
+     * // = Some(0.75) (correct)
+     * List.of(1.0, 10e100, 2.0, -10e100).average()
+     * }</pre>
+     *
      * @return {@code Some(average)} or {@code None}, if there are no elements
      * @throws UnsupportedOperationException if this elements are not numeric
      */
-    @SuppressWarnings({ "unchecked", "OptionalGetWithoutIsPresent" })
     default Option<Double> average() {
-        if (isEmpty()) {
-            return Option.none();
-        } else {
-            final Traversable<?> objects = isTraversableAgain() ? this : toStream();
-            final Object o = objects.head();
-            if (o instanceof Number) {
-                final Traversable<Number> numbers = (Traversable<Number>) objects;
-                final double d;
-                if (o instanceof Integer || o instanceof Long || o instanceof Byte || o instanceof BigInteger || o instanceof Short) {
-                    d = numbers.toJavaStream()
-                            .mapToLong(Number::longValue)
-                            .average()
-                            .getAsDouble();
+        try {
+            @SuppressWarnings("unchecked")
+            final double[] sum = ((Traversable<Number>) this).map(Number::doubleValue)
+                    .foldLeft(new double[4], (acc, d) -> { // acc = [sum, c, count, simpleSum]
+                        final double tmp = acc[0] + d;
+                        acc[1] += (Math.abs(acc[0]) >= Math.abs(d)) ? (acc[0] - tmp) + d : (d - tmp) + acc[0];
+                        acc[0] = tmp;
+                        acc[2]++;
+                        acc[3] += d;
+                        return acc;
+                    });
+            final double count = sum[2];
+            if (count == 0) {
+                return Option.none();
+            } else {
+                final double avg = (sum[0] + sum[1]) / count;
```

Comment:
Not that I terribly care, but why declare everything `final`? Are you trying have the scala equivalent of `val` ?

Reply:
Yes, this is specific for our internal Vavr implementation (and was part of our contribution guidelines I think but I can't find it anymore).

The only thing that can be remarked about it is that it is verbose. But in my experience programs get more robust.

The argument against it I hear most often is: "But the object of this variable isn't immutable, I can still mutate it." - and I say: "Yes, but the semantics of final is different, the reference to that object is immutable."

Long story short, it leads to more robust programs because it enforces the developer to not use variables if possible.

It does not mean to not use mutable variables at all.